frontends/p4/typeChecking/typeUnification.cpp

```diff
@@ -119,15 +119,16 @@ bool TypeUnification::unifyFunctions(const IR::Node* errorPosition,
     }
     if (src->returnType != nullptr)
         constraints->addEqualityConstraint(dest->returnType, src->returnType);
-    if (dest->parameters->size() != src->parameters->size()) {
-        if (reportErrors)
-            ::error("%1%: Cannot unify functions with different number of arguments: %2% to %3%",
-                    errorPosition, src, dest);
-        return false;
-    }
 
     auto sit = src->parameters->parameters.begin();
     for (auto dit : *dest->parameters->getEnumerator()) {
+        if (sit == src->parameters->parameters.end()) {
+            if (dit->getAnnotation("optional"))
```

Comment:
this should increment the sit if the source parameter is optional. Otherwise, you handle optional parameters only at the end of the list.